src/Icicle/Avalanche/Statement/Simp/Melt.hs

```diff
@@ -21,35 +23,88 @@ import              P
 import qualified    Data.Map            as Map
 
 
-pattern PrimZip  ta tb = PrimArray   (PrimArrayZip ta tb)
-pattern PrimPair ta tb = PrimMinimal (Min.PrimConst (Min.PrimConstPair ta tb))
-pattern PrimFst  ta tb = PrimMinimal (Min.PrimPair  (Min.PrimPairFst   ta tb))
-pattern PrimSnd  ta tb = PrimMinimal (Min.PrimPair  (Min.PrimPairSnd   ta tb))
-pattern PrimMkOpt   tv = PrimOption  (PrimOptionPack          tv)
-pattern PrimIsSome  tv = PrimProject (PrimProjectOptionIsSome tv)
-pattern PrimGet     tv = PrimUnsafe  (PrimUnsafeOptionGet     tv)
+------------------------------------------------------------------------
 
+-- this could be shared between modules if it's useful?
 
-melt :: (Show n, Ord n)
-     => a
-     -> Statement a n Prim
-     -> Fresh n (Statement a n Prim)
-melt a_fresh statements
- = transformUDStmt goStmt Map.empty statements
+pattern PrimZip     ta tb = PrimArray   (PrimArrayZip ta tb)
+pattern PrimPair    ta tb = PrimMinimal (Min.PrimConst (Min.PrimConstPair ta tb))
+pattern PrimFst     ta tb = PrimMinimal (Min.PrimPair  (Min.PrimPairFst   ta tb))
+pattern PrimSnd     ta tb = PrimMinimal (Min.PrimPair  (Min.PrimPairSnd   ta tb))
+
+pattern PrimMkOpt      tv = PrimPack    (PrimOptionPack          tv)
+pattern PrimIsSome     tv = PrimProject (PrimProjectOptionIsSome tv)
+pattern PrimGet        tv = PrimUnsafe  (PrimUnsafeOptionGet     tv)
+
+pattern PrimMkSum   ta tb = PrimPack    (PrimSumPack           ta tb)
+pattern PrimIsRight ta tb = PrimProject (PrimProjectSumIsRight ta tb)
+pattern PrimLeft    ta tb = PrimUnsafe  (PrimUnsafeSumGetLeft  ta tb)
+pattern PrimRight   ta tb = PrimUnsafe  (PrimUnsafeSumGetRight ta tb)
+
+data MeltOps a n p = MeltOps {
+    xPrim  :: p -> Exp a n p
+  , xVar   :: Name n -> Exp a n p
+  , xValue :: ValType   -> BaseValue -> Exp a n p
+  , xApp   :: Exp a n p -> Exp a n p -> Exp a n p
+
+  , primZip  :: ValType -> ValType -> Name n -> Name n -> Exp a n p
+  , primPair :: ValType -> ValType -> Name n -> Name n -> Exp a n p
+  , primFst  :: ValType -> ValType -> Exp a n p        -> Exp a n p
+  , primSnd  :: ValType -> ValType -> Exp a n p        -> Exp a n p
+
+  , primMkOpt  :: ValType -> Name n -> Name n -> Exp a n p
+  , primIsSome :: ValType -> Exp a n p        -> Exp a n p
+  , primGet    :: ValType -> Exp a n p        -> Exp a n p
+
+  , primMkSum   :: ValType -> ValType -> Name n -> Name n -> Name n -> Exp a n p
+  , primIsRight :: ValType -> ValType -> Exp a n p                  -> Exp a n p
+  , primLeft    :: ValType -> ValType -> Exp a n p                  -> Exp a n p
+  , primRight   :: ValType -> ValType -> Exp a n p                  -> Exp a n p
+  }
```

Comment:
I'm thinking I might move them to `Icicle.Avalanche.Combinators` in a separate PR?